File: C/TCP-examples/tcp-1-server.cc
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <arpa/inet.h>
#include <sys/errno.h>
#include <string.h>
#include "support.h"

//
// Sample TCP server. This simply reads lines of text from the remote
// connection and prints them out in the server.
//

const char *me = "SERVER";

int
main(int argc, char **argv)
{
  struct sockaddr_in addr;
  int listenfd;

  fprintf(stdout, "%s: Specified network is *%s*\n", me, argv[1]);
  fprintf(stdout, "%s: Specified port is *%s*\n", me, argv[2]);


  listenfd = socket(AF_INET, SOCK_STREAM, 0);
  check_and_exit(listenfd, "socket");
  //
  // Initialize socket
  //
  memset(&addr, 0, sizeof(addr));
  addr.sin_family = AF_INET;
  inet_pton(AF_INET, argv[1], &addr.sin_addr);
  addr.sin_port = htons(atoi(argv[2]));

  //
  // Bind socket to specified address
  //
  int ret = bind(listenfd, (struct sockaddr *)&addr, sizeof(addr));
  check_and_exit(ret, "bind");

  //
  // Place socket in LISTEN state - this is how the O/S know's
  // we're a server.
  //
  ret = listen(listenfd, 100);
  check_and_exit(ret,"listen");

  //
  // Start the acceptance loop
  //
  for(;;) {
    struct sockaddr_in theiraddr;
    socklen_t theiraddr_len = sizeof(theiraddr);
    int acceptfd = accept(listenfd, 
			  (struct sockaddr *)&theiraddr,
			  &theiraddr_len);
    check_and_exit(acceptfd, "accept");
    
    //
    // Describe the end points
    //
    print_socket_endpoints(stderr,"listen", listenfd);
    print_socket_endpoints(stderr,"accept", acceptfd);
   
    //
    // Begin file I/O on this socket
    //
    for(;;) {
      char line[1024];
      ssize_t r = read(acceptfd, line, sizeof(line));
      if ( r < 0 ) {
	perror("read");
	break;
      } else  if ( r ==  0 ) {
	fprintf(stderr,"%s: ** End of input, exit\n", me);
	break;
      } else {
	fprintf(stderr,"%s:** Got %d bytes: %s\n", me, (int) r, line);
      }
    }
    //
    // Close socket, wait for another connection
    //
    close(acceptfd);
  }
}

File: Python/TCP-examples/tcp-1-server.py
#!/usr/bin/env python3
##
## Corresponds to tcp-1-server.c
##
import socket
import sys,os


##
## Sample TCP server - receives messages from a client program
## but does not send data to the client.
##

if len(sys.argv) < 3:
    print("usage:", sys.argv[0], "<network> <port>")
    sys.exit(1)

network = sys.argv[1]
port = sys.argv[2]
me = "server"

##
## Print socket information
##
def print_socket_endpoints(name, sockfd):
    print("%s: Socket %s is connecting %s(me) and %s(them)" % (
        me, name, sockfd.getsockname(), sockfd.getpeername() ))

##
## Handle an interacton with a client
##
def handle_client(sockfd):
    while True:
        line = acceptfd.recv(8192)
        if len(line) == 0:
            print("End of input, exit")
            break
        print("%s: received %s" % (me, line.decode('utf-8')))
    

##
## Create a socket, bind to any interface, specific port
##
listenfd = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sendAddress = (network, int(port))
listenfd.bind(sendAddress)
listenfd.listen(100)

##
## Accept connections forever
##
while True:
    acceptfd, theirAddr = listenfd.accept()
    print_socket_endpoints("listen", acceptfd)
    handle_client(acceptfd)
    acceptfd.close()
print("Done")
